### DISCUSSION TITLE
Best way to make XHR requests for Header, Footer and SideNav components

### DISCUSSION THREAD

**Discussion by vasanthi-d** [vercel/next.js#37081] [Help]
Hi All,

My team is new to Next.js and we are just getting started with Next.js 12.x.  We have Header, Footer and SideNav that we need to display on all the pages. From next.js docs, I understand that we can use 'layouts' to achieve this. 
In our case, we need to make 
1st XHR request to get the markup for Header
2nd XHR request to get the markup for Footer
3rd XHR request for get the required js files to get the Header and Footer work properly.

End points for these XHR requests are known ahead of time and do not change. We can fetch this data only once and use it throughout the app.

I would like to have this data fetching done during build time using getStaticProps. But, it appears I cannot use getStaticProps in 'layouts' and can be used only Pages.  However, I do not want to repeat this code in every page and prefer to have in a common place.

What is the best way to achieve this?

**Comment by GabenGar** [comment]
If you are using the layout from docs you can pass arguments to it in the `_app` file. But you'll have to fetch the data in the `getStaticProps()` on each page for it to work. 
If it's known ahead of time, then you can just fetch the data in jsons in a separate script and import them into related components. Otherwise you are going to perform 3+ fetches per page built, which adds up pretty fast once you factor in multilang support.

**Reply by vasanthi-d** [reply]
@GabenGar Thank you for your response. Is there any recommendation on placement of the script that fetches the data ? It has to be executed as one of the bootstrapping scripts. Should we place it in _app.tsx? Can we have it as a react hook and use it different pages?

**Reply by GabenGar** [reply]
Neither, it has to be written as a completely separate script, outside of nextjs flow, which fetches the data and and saves it into json files. Probably written in vanilla nodejs, because nextjs abstracts a lot of things which crawl back when you write scripts like these.
A good idea is to have let's say `scripts` folder in the root of your project which would store all scripts like this. Let's say you create `fetch-layout-meta.js` module within it. Then you add `"npm run layout": "node scripts/fetch-layout-meta.js"` to your `scripts` section within `package.json` file.
Then just type `npm run layout` to refetch it if needed.

**Reply by vasanthi-d** [reply]
Thank you so much for your response. I'll give it a try and post the update.
